src/memory/host-tlb.c

```diff
@@ -86,46 +178,43 @@ static void hosttlb_write_slowpath(struct Decode *s, vaddr_t vaddr, int len, wor
   paddr_t paddr = va2pa(s, vaddr, len, MEM_TYPE_WRITE);
   paddr_write(paddr, len, data, cpu.mode, vaddr);
   if (likely(in_pmem(paddr))) {
-    HostTLBEntry *e = &hostwtlb[hosttlb_idx(vaddr)];
-    e->offset = guest_to_host(paddr) - vaddr;
-    e->gvpn = hosttlb_vpn(vaddr);
+    hosttlb_insert(vaddr, paddr, MEM_TYPE_WRITE);
   }
 }
 
 word_t hosttlb_read(struct Decode *s, vaddr_t vaddr, int len, int type) {
   Logm("hosttlb_reading " FMT_WORD, vaddr);
 #ifdef CONFIG_RVH
   extern bool has_two_stage_translation();
-  if(has_two_stage_translation()){
+  if (has_two_stage_translation()) {
     paddr_t paddr = va2pa(s, vaddr, len, type);
     return paddr_read(paddr, len, type, cpu.mode, vaddr);
   }
 #endif
-  vaddr_t gvpn = hosttlb_vpn(vaddr);
-  HostTLBEntry *e = type == MEM_TYPE_IFETCH ?
-    &hostxtlb[hosttlb_idx(vaddr)] : &hostrtlb[hosttlb_idx(vaddr)];
-  if (unlikely(e->gvpn != gvpn)) {
+  uint8_t *dst_ptr = hosttlb_lookup(vaddr, type);
+  if (unlikely(dst_ptr == HOSTTLB_PTR_FAIL_RET)) {
     Logm("Host TLB slow path");
     return hosttlb_read_slowpath(s, vaddr, len, type);
   } else {
     Logm("Host TLB fast path");
-    return host_read(e->offset + vaddr, len);
+    return host_read(dst_ptr, len);
   }
 }
 
 void hosttlb_write(struct Decode *s, vaddr_t vaddr, int len, word_t data) {
-  #ifdef CONFIG_RVH
+#ifdef CONFIG_RVH
   extern bool has_two_stage_translation();
-  if(has_two_stage_translation()){
+  if (has_two_stage_translation()) { 
```

Comment:
The reason why I put two stage translation in this place is to skip tlb translation when NEMU don't have tlb for the second stage. If you add the second stage tlb to NEMU, you should change these codes.

Reply:
But I only add the TLB for the second stage, which means it still need to skip previous host tlb lookup(va -> pa). 

P.S.
Maybe I can add a TLB mapping GVA to GPA. However, this cannot avoid adding judgement here, and asks page table walker(ptw) to be more granular. I used to do that, but it seems quiet difficult to keep it from a number of macros. Maybe when everything has been done, a code review can be applied.